lib/vnet/local_app_provider.go

```diff
@@ -0,0 +1,216 @@
+	profileNames, err := p.ClientApplication.ListProfiles()
+	if err != nil {
+		return nil, trace.Wrap(err, "listing profiles")
+	}
+	for _, profileName := range profileNames {
+		if fqdn == fullyQualify(profileName) {
+			// This is a query for the proxy address, which we'll never want to handle.
+			return nil, errNoTCPHandler
+		}
+
+		clusterClient, err := p.clusterClientForAppFQDN(ctx, profileName, fqdn)
+		if err != nil {
+			if errors.Is(err, errNoMatch) {
+				continue
+			}
+			// The user might be logged out from this one cluster (and retryWithRelogin isn't working). Log
+			// the error but don't return it so that DNS resolution will be forwarded upstream instead of
+			// failing, to avoid breaking e.g. web app access (we don't know if this is a web or TCP app yet
+			// because we can't log in).
+			log.ErrorContext(ctx, "Failed to get teleport client.", "error", err)
+			continue
+		}
+
+		leafClusterName := ""
+		clusterName := clusterClient.ClusterName()
+		if clusterName != "" && clusterName != clusterClient.RootClusterName() {
+			leafClusterName = clusterName
+		}
+
+		return p.resolveAppInfoForCluster(ctx, clusterClient, profileName, leafClusterName, fqdn)
+	}
+	// fqdn did not match any profile, forward the request upstream.
+	return nil, errNoTCPHandler
+}
+
+func (p *localAppProvider) clusterClientForAppFQDN(ctx context.Context, profileName, fqdn string) (ClusterClient, error) {
+	rootClient, err := p.ClientApplication.GetCachedClient(ctx, profileName, "")
+	if err != nil {
+		log.ErrorContext(ctx, "Failed to get root cluster client, apps in this cluster will not be resolved.", "profile", profileName, "error", err)
+		return nil, errNoMatch
+	}
+
+	if isDescendantSubdomain(fqdn, profileName) {
+		// The queried app fqdn is a subdomain of this cluster proxy address.
+		return rootClient, nil
+	}
+
+	leafClusters, err := getLeafClusters(ctx, rootClient)
+	if err != nil {
+		// Good chance we're here because the user is not logged in to the profile.
+		log.ErrorContext(ctx, "Failed to list leaf clusters, apps in this cluster will not be resolved.", "profile", profileName, "error", err)
+		return nil, errNoMatch
+	}
+
+	// Prefix with an empty string to represent the root cluster.
+	allClusters := append([]string{""}, leafClusters...)
+	for _, leafClusterName := range allClusters {
+		clusterClient, err := p.ClientApplication.GetCachedClient(ctx, profileName, leafClusterName)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to get cluster client, apps in this cluster will not be resolved.", "profile", profileName, "leaf_cluster", leafClusterName, "error", err)
+			continue
+		}
+
+		clusterConfig, err := p.clusterConfigCache.GetClusterConfig(ctx, clusterClient)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to get VnetConfig, apps in the cluster will not be resolved.", "profile", profileName, "leaf_cluster", leafClusterName, "error", err)
```

Comment:
```suggestion
			log.ErrorContext(ctx, "Failed to get VNet config, apps in this cluster will not be resolved.", "profile", profileName, "leaf_cluster", leafClusterName, "error", err)
```

Just to align this with the log message on line 134.